src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -28928,6 +28928,274 @@ public void Init(T t)
+            // It's not clear whether diagnostic 1 and 2 are appropriate.
+            // Going by the principle of "if you put a good value in, you will get a good value out",
+            // it seems like property state should be MaybeNull after assigning a MaybeDefault to the property.
+            // https://github.com/dotnet/roslyn/issues/49964
+
+            // Note that the lack of constructor exit warning in 'C()' is expected due to presence of 'AllowNull' on the property declaration.
+            comp.VerifyDiagnostics(
+                // (12,12): warning CS8604: Possible null reference argument for parameter 't' in 'void C<T>.M1(T t)'.
+                //         M1(Prop); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "Prop").WithArguments("t", "void C<T>.M1(T t)").WithLocation(12, 12),
+                // (18,12): warning CS8604: Possible null reference argument for parameter 't' in 'void C<T>.M1(T t)'.
+                //         M1(Prop); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "Prop").WithArguments("t", "void C<T>.M1(T t)").WithLocation(18, 12),
+                // (19,9): warning CS8602: Dereference of a possibly null reference.
+                //         Prop.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "Prop").WithLocation(19, 9));
+        }
+
+        [Fact, WorkItem(49964, "https://github.com/dotnet/roslyn/issues/49964")]
+        public void AdjustPropertyState_Generic_MaybeNullInput_NotNullOutput()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public class C<T> {
+    [NotNull]
+    T Prop { get; set; }
+
+    public C() // 1
+    {
+    }
+
+    public C(T t) // 2
+    {
+        Prop = t;
+    }
+
+    public void M(T t) {
+        Prop = t;
+        Prop.ToString(); // 3
+    }
+
+    public void M1(T t) { }
+}
+";
+            var comp = CreateNullableCompilation(new[] { source, NotNullAttributeDefinition });
+            // It's not clear whether diagnostic 2 or 3 are appropriate.
+            // Going by the principle of "if you put a good value in, you will get a good value out",
+            // it seems like property state should be NotNull after assigning a MaybeNull to the property.
+            // https://github.com/dotnet/roslyn/issues/49964
+            comp.VerifyDiagnostics(
+                // (8,12): warning CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
+                //     public C() // 1
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("property", "Prop").WithLocation(8, 12),
+                // (12,12): warning CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
+                //     public C(T t) // 2
```

Comment:
>public C(T t) // 2 [](start = 23, length = 18)

Is this a new warning? If so, we should consider investigating #49964 first, to make sure it is appropriate.

Reply:
No, the same warnings are given before this change. [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABABgqYEYA6AEQEsIAc1wB7WHwDGULgGERKOAEEahAJ5Q+UANyNGTAMzsATBRkAeACoA+CgG9GFRxQDaAORExXRQgF0HTiwoABQQRAAc7CkE4GC0KKBi4gF9dBid2QxkACgBKCgB6fIoOf0d7NKcUhlKM0yzAmDzCiiMa8vT0kPCKAF4KWJqqmoN2ABYKAFl6/rz2jscuiL6BivnFrgsRAGUYBD5cQVy45v1B1PSRpnGJjmnGyKqkoA===)